lib: remove queue implementation from JSStreamWrap

addaleax opened the pull request:
The streams implementation generally ensures that only one write()
call is active at a time. `JSStreamWrap` instances still kept
queue of write reqeuests in spite of that; refactor it away.

Also, fold `isAlive()` into a constant function on the native side.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)

lib

bnoordhuis commented on lib/internal/wrap_js_stream.js:
```diff
@@ -60,7 +66,10 @@ class JSStreamWrap extends Socket {
+    this[kCurrentWriteRequest] = null;
+    this[kCurrentShutdownRequest] = null;
+
+    // Start reading
```
Dot at end.  Also, somewhat superfluous comment.

addaleax replied:
I don’t think it’s superfluous – I could imagine there’s quite a few who wouldn’t know why you’d ever want to call `.read(0)`.

bnoordhuis commented on lib/internal/wrap_js_stream.js:
```diff
@@ -88,33 +93,56 @@ class JSStreamWrap extends Socket {
     this.stream.cork();
-    for (var n = 0; n < bufs.length; n++)
-      this.stream.write(bufs[n], done);
+    for (var i = 0; i < bufs.length; i++)
```
If you're making stylistic changes anyway: `let i = ...` (or `for (const buf of bufs)`)

addaleax replied:
We are not allowed to use `let` in loops, the linter yells about that. :(

`for (const buf of bufs)` seems to work though!

bnoordhuis commented on lib/internal/wrap_js_stream.js:
```diff
@@ -88,33 +93,56 @@ class JSStreamWrap extends Socket {
+    this[kCurrentWriteRequest] = req;
 
-    var pending = bufs.length;
+    const handle = this._handle;
```
Unused?

addaleax replied:
The linter would complain about that. :)

bnoordhuis replied:
I'm not seeing it, though.  Same for the `const handle = ...` on line 112.

addaleax replied:
It’s passed to the `self.finishWrite()` call down on line 162

bnoordhuis replied:
Oh, indeed.  Unfortunate fold in the diff.

jasnell commented on lib/internal/wrap_js_stream.js:
```diff
@@ -27,12 +30,15 @@ class JSStreamWrap extends Socket {
-    handle.onwrite = (req, bufs) => this.doWrite(req, bufs);
+    // Inside of the following functions, `this` refers to the handle
+    // and `this.owner` refers to this JSStreamWrap instance.
+    handle.isClosing = function() { return this.owner.isClosing(); };
```
would be slightly better to avoid the closures entirely and just make these top level functions.

addaleax replied:
@jasnell done!

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-commit/15175/

mcollina commented on lib/internal/wrap_js_stream.js:
```diff
@@ -88,33 +97,56 @@ class JSStreamWrap extends Socket {
     this.stream.cork();
-    for (var n = 0; n < bufs.length; n++)
-      this.stream.write(bufs[n], done);
+    for (const buf of bufs)
```
can you use a `for(;;)` loop here? for-of is still extremely slower than `for(;;)`.

mcollina commented on lib/internal/wrap_js_stream.js:
```diff
@@ -126,93 +158,42 @@ class JSStreamWrap extends Socket {
-
-        handle.finishWrite(req, errCode);
+      setImmediate(() => {
+        self.finishWrite(handle, errCode);
```
you can avoid this closure completely with `setImmediate(finishWrite, self, handle, errCode)`.

mcollina commented on lib/internal/wrap_js_stream.js:
```diff
@@ -126,93 +158,42 @@ class JSStreamWrap extends Socket {
+
+      this.finishWrite(handle, uv.UV_ECANCELED);
+      this.finishShutdown(handle, uv.UV_ECANCELED);
+
```
can you remove this closure as well?

addaleax commented:
@mcollina I’ll get to addressing your comments later, but I’ll point out that at least in the case of the `for` loop that was a suggestion by an earlier comment (I can switch back tho) and that this *is* the slow path already, so squeezing the last bits of performance out of it might not be that important

addaleax commented on lib/internal/wrap_js_stream.js:
```diff
@@ -88,33 +97,56 @@ class JSStreamWrap extends Socket {
 
-    var pending = bufs.length;
+    const handle = this._handle;
+    const self = this;
```
@billouboq it’s used inside the `done()` function below :)

addaleax commented:
@mcollina I’ve addressed the `for (const … of …)` suggestion by you.

I also tried to benchmark all of the suggested changes via the benchmark added in https://github.com/nodejs/node/pull/17983, but did not get a statistically significant improvement as a result, so I’d prefer to stick to the more readable code here.

mcollina commented:
I'm 👍  if you want to keep if it does not regress anything. I prefer to not use code that might become problematic just for stylistic reasons (and `for (;;)` is beautiful).

Have you thought about refactoring all those closures as well? I don't expect to get any perf boost, but it might help memory consumption and the gc in the long run. Feel free to ignore those, it's just a nit.

addaleax commented:
@mcollina I benchmarked all of your suggestions together (via that linked benchmark). I hadn’t really thought about gc, thanks for the reminder! What would be the best way to measure the impact of a change like that?

bnoordhuis commented:
@addaleax Slight tangent but an idea I've had is roughly to:

1. Instrument with `v8::Isolate::AddGCEpilogueCallback()`
2. Log memory usage reported by `v8::Isolate::GetHeapSpaceStatistics()` after every GC
3. Collate the data after the process exits

With the idea that you compare runs of the benchmarks or the test suite to see if there are large swings in GC frequency or heap usage.

mcollina commented:
I used `--trace-gc-object-stats` in the past with great success.

joyeecheung commented:
@bnoordhuis I think what you've descried is basically what `--trace_gc --trace_gc_verbose` does?

bnoordhuis commented:
By 'log' I mean 'write to file' in a format that's machine readable.

`--trace_gc` dumps everything to stdout interleaved with normal output.  Okay for eyeballing stuff, not so great for automation.

addaleax commented:
Landed in b171adc4d12e96f7b72155ea54a78dd1c8e5823f

I’ll take a look at the closures at another point, this PR isn’t introducing new ones and I’d like to get around to addressing the issue from https://github.com/nodejs/node/pull/17938#issuecomment-354683850

MylesBorins commented:
This is currently breaking some stuff on v9.x

It is possible that after https://github.com/nodejs/node/pull/18050 this will work

addaleax commented:
@MylesBorins It applies and tests cleanly on my machine … could you show the error you were seeing?

MylesBorins commented:
@addaleax I'm assuming it was unblocked by the http2 update. Will look at this again in a bit
